driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2002,20 +2002,38 @@ public void refreshSchemaAndSignal(final Connection connection, final DefaultRes
             if (logger.isDebugEnabled())
                 logger.debug("Refreshing schema for {}{}", keyspace == null ? "" : keyspace, table == null ? "" : '.' + table);
 
+            maybeRefreshSchemaAndSignal(connection, future, rs, keyspace, table);
+        }
+
+        public void waitForSchemaAgreementAndSignal(final Connection connection, final DefaultResultSetFuture future, final ResultSet rs) {
+            maybeRefreshSchemaAndSignal(connection, future, rs, null, null);
+        }
+
+        private void maybeRefreshSchemaAndSignal(final Connection connection, final DefaultResultSetFuture future, final ResultSet rs, final String keyspace, final String table) {
+            final boolean refreshSchema = (keyspace != null); // if false, only wait for schema agreement
+
             executor.submit(new Runnable() {
```

Comment:
On possible concern I have with this is that if you have multiple schema updates happening at a time, you could have multiple executor threads hung up in waiting for schema agreement (up to 10 seconds), but that doesn't seem very likely.   If you have low thread count on the executor this could defer host state updates. 

An idea to optimize this is that you could change this so you first submit a callable that waits for schema agreement.  This way other callers that enter this method could listen on that future and then depending on the result of the future react and then submit a Runnable that does the post tasks (refresh schema, etc.).  So if the future completes successfully those listening knows that the schema is in agreement.  That adds a bit of complexity and I don't know if it provides much value but thought I'd bring it up to see what you think.  Also this is how the code already was, so its not completed related to this pull request.